File: sprintboot-Security&JWT/jwt/src/main/java/com/cos/jwt/config/jwt/JwtAuthenticationFilter.java
package com.cos.jwt.config.jwt;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;

/** /login 요청해서 username,password 전송하면
 * UsernamePasswordAuthenticationFilter 동작을 함
 */
@RequiredArgsConstructor
public class JwtAuthenticationFilter extends UsernamePasswordAuthenticationFilter {

    private final AuthenticationManager authenticationManager;


    /** 로그인 요청을 하면 로그인 시도를 위해서 실행되는 함수 */
    @Override
    public Authentication attemptAuthentication(HttpServletRequest request, HttpServletResponse response) throws AuthenticationException {
        System.out.println("JwtAuthenticationFilter 로그인 시도중");

        return super.attemptAuthentication(request, response);
    }
}

File: kotlin-spring-security/demo/src/main/kotlin/com/example/demo/common/authority/JwtAuthenticationFilter.kt
package com.example.demo.common.authority

import jakarta.servlet.FilterChain
import jakarta.servlet.ServletRequest
import jakarta.servlet.ServletResponse
import jakarta.servlet.http.HttpServletRequest
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.util.StringUtils
import org.springframework.web.filter.GenericFilterBean

class JwtAuthenticationFilter(

    /** JWT 토큰
     *  JWT(Json Web Token)를 사용하여 Spring Security에서 인증된 사용자를 설정하는 필터 클래스
     *  doFilter 메서드에서 HTTP 요청에서 토큰을 추출하고,
     *  해당 토큰이 유효하면 Spring Security의 SecurityContextHolder에 사용자 정보를 설정
     */
    private val jwtTokenProvider: JwtTokenProvider
) : GenericFilterBean() {
    override fun doFilter(
        request: ServletRequest?,
        response: ServletResponse?,
        chain: FilterChain?
    ) {
        val token = resolveToken(request as HttpServletRequest)

        if (token != null && jwtTokenProvider.validateToken(token)) {
            val authentication = jwtTokenProvider.getAuthentication(token)
            SecurityContextHolder.getContext().authentication = authentication
        }

        chain?.doFilter(request, response)
    }

    // HTTP 요청에서 토큰을 추출하는 메서드
    private fun resolveToken(request: HttpServletRequest): String? {
        val bearerToken = request.getHeader("Authorization")
        return if (StringUtils.hasText(bearerToken) &&
            bearerToken.startsWith("Bearer")) {
            bearerToken.substring(7)
        } else {
            null
        }
    }
}
